### PR TITLE
Rework error handling

### PR DESCRIPTION
**Pull request by finestructure**
Rationale: https://noteplan.co/n/575ADC0D-8D6C-4ADF-BB25-972717C42C41

To do

- [x] Wait for upstream merge and release of https://github.com/pointfreeco/swift-concurrency-extras/pull/52, then remove dependency pin on fork
- [x] Write up a rationale for this change (based on https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/issues/2639 and the extended `test_ingest_continue_on_error`)
- [x] Let it run on staging for a few days
- [x] Check if we're seeing any weird error handling issues
  - [x] is the error formatting ok?
  - [x] do we see new errors popping up?
  - [x] are some errors not popping up anymore?
  - [x] cause some errors on `dev` to confirm things work as expected

This is best reviewed with the "ignore whitespace" setting to hide all the indentation noise from new namespaces.

### CODE REVIEW

File: Sources/App/Commands/Common.swift [hunk 1]
```diff
@@ -17,111 +17,125 @@ import PostgresKit
...
 
 
-/// Update packages (in the `[Result<Joined<Package, Repository>, Error>]` array).
-///
-/// - Parameters:
-///   - client: `Client` object
-///   - database: `Database` object
-///   - results: `Joined<Package, Repository>` results to update
-///   - stage: Processing stage
-func updatePackages(client: Client,
```
Comment (finestructure) [inline Sources/App/Commands/Common.swift:27]: `updatePackages` (note the plural) was only called from analysis. Moving this into the `Analyze` namespace (which I'll update to `Analysis` at some point in the future) makes this explicit.

File: Tests/AppTests/IngestorTests.swift [hunk 2]
```diff
@@ -83,12 +83,17 @@ class IngestorTests: AppTestCase {
...
-        XCTAssertEqual(repo.licenseUrl, "license")
+        do {
+            // validate the second package's license is updated
+            let repo = try await Repository.query(on: app.db)
+                .filter(\.$name == "2")
+                .first()
+                .unwrap()
+            XCTAssertEqual(repo.licenseUrl, "license")
+            for pkg in try await Package.query(on: app.db).all() {
+                XCTAssertEqual(pkg.processingStage, .ingestion, "\(pkg.url) must be in ingestion")
```
Comment (finestructure) [inline Tests/AppTests/IngestorTests.swift:94]: This is a crucial bit of additional test coverage where we ensure the processing stage is advanced when there are errors.

File: Tests/AppTests/IngestorTests.swift [hunk 3]
```diff
@@ -354,12 +364,12 @@ class IngestorTests: AppTestCase {
 
     func test_ingest_unique_owner_name_violation() async throws {
         // Test error behaviour when two packages resolving to the same owner/name are ingested:
-        //   - don't update package
         //   - don't create repository records
```
Comment (finestructure) [inline Tests/AppTests/IngestorTests.swift:367]: This is an important change to how we're dealing with this error scenario. Previously, we did two things:

- avoid creating a duplicate repository record
- avoid updating the related package

We could preserve this behaviour but it actually simplifies things quite a bit to drop the second part of this. And it's the theme of this change: make sure processing is advanced through a stages despite errors.

In order to ensure this change doesn't break things in the following stage, this test now explicitly runs analysis as part of its validation to ensure it doesn't raise any errors (see below).

File: Tests/AppTests/IngestorTests.swift [hunk 4]
```diff
@@ -390,31 +399,51 @@ class IngestorTests: AppTestCase {
...
+                if cmd.description.hasSuffix("package dump-package") {
+                    return .packageDump(name: "foo")
+                }
+                return ""
+            }
+
+            try await Analyze.analyze(client: app.client, database: db, mode: .id(.id0))
+            try await Analyze.analyze(client: app.client, database: db, mode: .id(.id1))
+            try await XCTAssertEqualAsync(try await Package.find(.id0, on: db)?.processingStage, .analysis)
+            try await XCTAssertEqualAsync(try await Package.find(.id1, on: db)?.processingStage, .analysis)
```
Comment (finestructure) [inline Tests/AppTests/IngestorTests.swift:446]: This is where we're ensuring analysis can deal with this scenario where one package doesn't have a corresponding repository record.

File: Sources/App/Core/Dependencies/EnvironmentClient.swift [hunk 5]
```diff
@@ -44,6 +44,16 @@ struct EnvironmentClient {
...
+
+    enum FailureMode: String {
+        case fetchMetadataFailed
+        case findOrCreateRepositoryFailed
+        case invalidURL
+        case noRepositoryMetadata
+        case repositorySaveFailed
+        case repositorySaveUniqueViolation
+    }
+    var shouldFail: @Sendable (_ failureMode: FailureMode) -> Bool = { _ in false }
```
Comment (finestructure) [inline Sources/App/Core/Dependencies/EnvironmentClient.swift:56]: I've added this error triggering mechanism, which allows us to configure an environment variable `FAILURE_MODE` as for example

```json
{
    "fetchMetadataFailed": 0.1,
    "findOrCreateRepositoryFailed": 0,
    "invalidURL": 0,
    "noRepositoryMetadata": 0,
    "repositorySaveFailed": 0,
    "repositorySaveUniqueViolation": 0
}
```

which will trigger a `fetchMetadataFailed` error at a rate of 10%.

I'm using this to ensure on `dev` that there's no knock-on effect from errors in subsequent stages. It's also made it easier to ensure locally that the error logs are ok.

We can revert this after testing.

**Comment by finestructure** [comment]
Ok, this is all working as expected:

Screenshot 2024-12-15 at 13 52 19

**Review by daveverwer — APPROVED** [review]

**Comment by finestructure** [comment]
Now that we've got 1.3.1 of swift-concurrency-extras I've rebased all the package manifest changes out of the PR.

**Pull request merged**